csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/Windows/WindowsProtectedMemoryAllocatorLLP64.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Runtime.InteropServices;
+using GoDaddy.Asherah.PlatformNative.LLP64.Windows;
+using GoDaddy.Asherah.PlatformNative.LLP64.Windows.Enums;
+using GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Libc;
+
+namespace GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Windows
+{
+    internal abstract class WindowsProtectedMemoryAllocatorLLP64 : IProtectedMemoryAllocator
+    {
+        protected const int ErrorNotLocked = 158;
+        protected static readonly IntPtr InvalidPointer = new IntPtr(-1);
+
+        public abstract IntPtr Alloc(ulong length);
+
+        public abstract void Free(IntPtr pointer, ulong length);
+
+        public void SetNoAccess(IntPtr pointer, ulong length)
+        {
+            length = AdjustLength(length);
+
+            if (!WindowsInterop.CryptProtectMemory(pointer, (UIntPtr)length, CryptProtectMemoryOptions.SAME_PROCESS))
+            {
+                var errno = Marshal.GetLastWin32Error();
+                throw new LibcOperationFailedException("CryptProtectMemory", 0L, errno);
+            }
+
+            UnlockMemory(pointer, length);
+        }
+
+        public void SetReadAccess(IntPtr pointer, ulong length)
+        {
+            length = AdjustLength(length);
+
+            LockMemory(pointer, length);
+
+            if (!WindowsInterop.CryptUnprotectMemory(pointer, (UIntPtr)length, CryptProtectMemoryOptions.SAME_PROCESS))
+            {
+                var errno = Marshal.GetLastWin32Error();
+                throw new LibcOperationFailedException("CryptUnprotectMemory", 0L, errno);
+            }
+        }
+
+        public void SetReadWriteAccess(IntPtr pointer, ulong length)
+        {
+            length = AdjustLength(length);
+
+            LockMemory(pointer, length);
+
+            if (!WindowsInterop.CryptUnprotectMemory(pointer, (UIntPtr)length, CryptProtectMemoryOptions.SAME_PROCESS))
+            {
+                var errno = Marshal.GetLastWin32Error();
+                throw new LibcOperationFailedException("CryptUnprotectMemory", 0L, errno);
+            }
+        }
+
+        public void ZeroMemory(IntPtr pointer, ulong length)
+        {
+            WindowsInterop.ZeroMemory(pointer, (UIntPtr)length);
+        }
+
+        protected void LockMemory(IntPtr pointer, ulong length)
+        {
+            if (!WindowsInterop.VirtualLock(pointer, (UIntPtr)length))
+            {
+                var errno = Marshal.GetLastWin32Error();
+                throw new LibcOperationFailedException("VirtualLock", 0L, errno);
```

Comment:
seems a bit wonky to raise a `LibcOperationFailedException` from the `GoDaddy.Asherah.SecureMemory.ProtectedMemoryImpl.Windows` namespace. Perhaps a new `WindowsInteropOperationFailedException` is in order?

Reply:
I added an exception for this, but I derived it from the LibcOperationFailedException because I don't want to change the contract for Windows platform.  We'd need to introduce a new base exception type to base the existing classes off of, and then do a contract breaking release.